src/Servers/Kestrel/Core/src/Internal/HttpConnection.cs

```diff
@@ -39,80 +37,77 @@ public HttpConnection(HttpConnectionContext context)
-                                    break;
-                                case HttpProtocols.Http2:
-                                    // _http2Connection must be initialized before yielding control to the transport thread,
-                                    // to prevent a race condition where _http2Connection.Abort() is called just as
-                                    // _http2Connection is about to be initialized.
-                                    requestProcessor = new Http2Connection(derivedContext);
-                                    _protocolSelectionState = ProtocolSelectionState.Selected;
-                                    break;
-                                case HttpProtocols.None:
-                                    // An error was already logged in SelectProtocol(), but we should close the connection.
-                                    Abort(new ConnectionAbortedException(CoreStrings.ProtocolSelectionFailed));
-                                    break;
-                                default:
-                                    // SelectProtocol() only returns Http1, Http2 or None.
-                                    throw new NotSupportedException($"{nameof(SelectProtocol)} returned something other than Http1, Http2 or None.");
-                            }
-
-                            _requestProcessor = requestProcessor;
-                        }
-                    }
-
-                    var closedRegistration = _context.ConnectionContext.ConnectionClosed.Register(state => ((HttpConnection)state).OnInputOrOutputCompleted(), this);
-
-                    // We don't care about callbacks once all requests are processed
-                    using (closedRegistration)
-                    {
-                        if (requestProcessor != null)
-                        {
-                            await requestProcessor.ProcessRequestsAsync(httpApplication);
-                        }
-                    }
+                    var connectionHeartbeatFeature = _context.ConnectionFeatures.Get<IConnectionHeartbeatFeature>();
+                    var connectionLifetimeNotificationFeature = _context.ConnectionFeatures.Get<IConnectionLifetimeNotificationFeature>();
+
+                    // These features should never be null in Kestrel itself, if this middleware is ever refactored to run outside of kestrel,
+                    // we'll need to handle these missing.
+                    Debug.Assert(connectionHeartbeatFeature != null, nameof(IConnectionHeartbeatFeature) + " is missing!");
+                    Debug.Assert(connectionLifetimeNotificationFeature != null, nameof(IConnectionLifetimeNotificationFeature) + " is missing!");
+
+                    // Register the various callbacks once we're going to start processing requests
+
+                    // The heart beat for various timeouts
+                    connectionHeartbeatFeature?.OnHeartbeat(state => ((HttpConnection)state).Tick(), this);
```

Comment:
Why can't we remove these debug asserts and nullables? Arguably it is better to throw an NRE than let the server continue to run if these are null.

Reply:
Throw an NRE? It just means you don't get timeouts. That would be another server implementation so it's not a huge deal ATM (we're not there yet).

Reply:
Also, it's super unfortunate we didn't return `IDisposable` from this...